### PR TITLE
Allow multiple instances (using message broker)

### PR DESCRIPTION
**Pull request by isocolsky**

### CODE REVIEW

File: bws.js [hunk 1]
```diff
@@ -37,7 +37,7 @@ var start = function(cb) {
         if (err) return cb(err);
         var server = config.https ? serverModule.createServer(serverOpts, app) :
           serverModule.Server(app);
-        WsApp.start(server, config);
+        new WsApp().start(server, config);
```
Comment (matiu) [inline bws.js]: Just to simplify reading, I suggest `(new X()).start` or 
```
var x=new X();
x.start(
```

File: lib/blockchainmonitor.js [hunk 2]
```diff
@@ -5,100 +5,107 @@ var _ = require('lodash');
...
 
+  var self = this;
+
   var explorer = new BlockchainExplorer({
     provider: provider,
     network: network,
+    url: url,
   });
 
   var socket = explorer.initSocket();
```
Comment (matiu) [inline lib/blockchainmonitor.js]: It should die if the connection cannot be established.

File: lib/blockchainmonitor.js [hunk 3]
```diff
@@ -5,100 +5,107 @@ var _ = require('lodash');
...
-      data: {
-        address: address,
-        txid: txid,
-      },
-    });
-    self.emit('notification', notification);
-  };
+  var outs = _.compact(_.map(data.vout, function(v) {
+    var addr = _.keys(v)[0];
+    if (addr.indexOf('3') != 0 && addr.indexOf('2') != 0) return;
```
Comment (matiu) [inline lib/blockchainmonitor.js]: small opt:  
```
 startChar = addr.charAt(0);
 if (startChar != '2' && startChar != '3') return;
```

File: lib/blockchainmonitor.js [hunk 4]
```diff
@@ -5,100 +5,107 @@ var _ = require('lodash');
...
-  var subscriber = self.subscriber[network];
-  _.each(addresses, function(address) {
-    self.subscriptions[walletId].addresses.push(address);
-    subscriber.subscribe(address, _.bind(handlerFor, walletId, address));
+  }));
+  if (_.isEmpty(outs)) return;
+
+  async.each(outs, function(out, next) {
+    self.storage.fetchAddress(out.address, function(err, address) {
+      if (err || !address) return next(err);
```
Comment (matiu) [inline lib/blockchainmonitor.js]: if err we should die; or if the mongo driver try to reconnect itself, we should log.

File: lib/server.js [hunk 5]
```diff
@@ -282,22 +304,13 @@ WalletService.prototype._verifySignature = function(text, signature, pubKey) {
...
-
-/**
  * _notify
  *
  * @param {String} type
  * @param {Object} data
  * @param {Boolean} isGlobal - If true, the notification is not issued on behalf of any particular copayer (defaults to false)
  */
-WalletService.prototype._notify = function(type, data, isGlobal) {
+WalletService.prototype._notify = function(type, data, isGlobal, cb) {
```
Comment (matiu) [inline lib/server.js]: Maybe we should change to `type, data, opts, cb`

File: lib/storage.js [hunk 6]
```diff
@@ -312,6 +312,20 @@ Storage.prototype.storeAddressAndWallet = function(wallet, addresses, cb) {
   });
 };
 
+Storage.prototype.fetchAddress = function(address, cb) {
+  var self = this;
+
+  this.db.collection(collections.ADDRESSES).findOne({
+    address: address,
```
Comment (matiu) [inline lib/storage.js]: Are we using indexes for Address? `ensureIndex` (http://docs.mongodb.org/manual/reference/method/db.collection.ensureIndex/)

**Comment by matiu** [comment]
ACK. Great job!.

 Left some minor comments.

**Comment by isocolsky** [comment]
All issues addressed. Thanks for the review!

**Pull request merged**